rust/src/builder.rs

```diff
@@ -15,158 +15,97 @@
-        assert!(!self.data.is_null(), "cannot push onto uninitialized data");
-        if self.len == self.capacity {
-            // grow capacity by 64 bytes or double the current capacity, whichever is greater
-            let new_capacity = cmp::max(64, self.capacity * 2);
-            self.grow(new_capacity);
-        }
-        assert!(self.len < self.capacity, "new length exceeds capacity");
-        unsafe {
-            *self.data.offset(self.len as isize) = v;
-        }
-        self.len += 1;
-    }
-
-    /// Set a value at a slot in the allocated memory without adjusting the length
-    pub fn set(&mut self, i: usize, v: T) {
-        assert!(
-            !self.data.is_null(),
-            "cannot set value if data is uninitialized"
-        );
-        assert!(i < self.capacity, "index exceeds capacity");
-        unsafe {
-            *self.data.offset(i as isize) = v;
-        }
+    pub fn push(&mut self, v: T) -> Result<()> {
+        self.reserve(1)?;
+        self.write_bytes(v.to_byte_slice(), 1)
     }
 
-    /// push a slice of type T, growing the internal buffer as needed
-    pub fn push_slice(&mut self, slice: &[T]) {
-        self.reserve(slice.len());
-        let sz = mem::size_of::<T>();
-        unsafe {
-            libc::memcpy(
-                self.data.offset(self.len() as isize) as *mut libc::c_void,
-                slice.as_ptr() as *const libc::c_void,
-                slice.len() * sz,
-            );
-        }
-        self.len += slice.len();
+    /// Push a slice of type T, growing the internal buffer as needed
+    pub fn push_slice(&mut self, slice: &[T]) -> Result<()> {
+        let array_slots = slice.len() as i64;
+        self.reserve(array_slots)?;
+        self.write_bytes(slice.to_byte_slice(), array_slots)
     }
 
     /// Reserve memory for n elements of type T
-    pub fn reserve(&mut self, n: usize) {
-        if self.len + n > self.capacity {
-            let new_capacity = cmp::max(self.capacity * 2, n);
-            self.grow(new_capacity);
+    pub fn reserve(&mut self, n: i64) -> Result<()> {
+        let new_capacity = self.len + n;
+        if new_capacity > self.capacity() {
+            return self.grow(new_capacity);
         }
+        Ok(())
     }
 
-    /// Grow the buffer to the new size n (number of elements of type T)
-    fn grow(&mut self, new_capacity: usize) {
-        let sz = mem::size_of::<T>();
-        let old_buffer = self.data;
-        let new_buffer = allocate_aligned((new_capacity * sz) as i64).unwrap();
-        unsafe {
-            libc::memcpy(
-                new_buffer as *mut libc::c_void,
-                old_buffer as *const libc::c_void,
-                self.len * sz,
-            );
-        }
-        self.capacity = new_capacity;
-        self.data = new_buffer as *mut T;
-        free_aligned(old_buffer as *const u8);
+    /// Grow the buffer to the new capacity of size n (where n is the number of elements of type T)
```

Comment:
nit: add backtick to the parameters (i.e., `n` and `T`).

Reply:
Can you explain please @sunchao , I'm not quite understanding this one?